### ISSUE TITLE
Service/provider architecture

### ISSUE THREAD

**Issue by rowanseymour** [nyaruka/goflow#733]
We currently have discussions around how we should be implementing airtime transfers in goflow (https://github.com/rapidpro/rapidpro/issues/997), what NLP/NLU should look like in goflow (https://github.com/rapidpro/rapidpro/issues/832), and whether goflow should be responsible for sending emails (https://github.com/nyaruka/goflow/issues/537).

One common thread in each of those is a need to not tie elements in the flow spec to specific third party APIs. For example:

- the `transfer_airtime` action shouldn't be tied to the [TransferTo](https://www.transfer-to.com) API
- the `send_email` action shouldn't be concerned about whether it's sending via SMTP or AWS SES
- whatever we implement for NLP/NLU shouldn't be tied to any particular 3rd party service

The direction we went with airtime transfers in the past was to say that `transferto` was an extension that makes a `transfer_airtime` action. The implementation for talking to the TransferTo API lives in that action and it reads its configuration from the environment. This approach has a couple of shortcomings:

- There can only be one implementation of the `transfer_airtime` action.
- The only place to pass the TransferTo configuration to the action is in the environment - which is serialized in the session JSON. Not ideal to have API keys and passwords etc in that JSON.

A different approach we've chatted about but not yet formalized would be making those things accessible to flow elements as "services" with simple interfaces, e.g.

- `send_email` does its template evaluation and localization to get the subject, body, and addresses.. and then calls an email service `send(addresses, subject, body)`
- `transfer_airtime` just has its list of countries and amounts and calls an airtime transfer service with `transfer(country, amount)`
- NLP is accessible to flow elements via a simple interface like `classify(input, model)` etc

And then the caller provides "providers" which implement those services. That raises some questions:

1. How does the caller provide them 
2. How do the flow elements access them
3. How does the provided service get the configuration it needs, e.g. an email service provided by mailroom would need to get an org_id and go find the email configuration stored on the org.

Since these things would be org specific it would seem the obvious place for them is the environment. Thus far `Engine` has been a global thing re-usable across all sessions, and `Environment` has been where everything org specific goes. So for example in the `send_email` action...

```go
env.EmailService().Send(addresses, subject, body)
```

but the problem with that is that environment's are currently (de)serializable. When the caller loads a session from JSON that includes the environment. A trigger is serializable with its environment.

So perhaps better approach is to make services accessible thru the engine, i.e.

```go
run.Session().Engine().EmailService().Send(addresses, subject, body)
```

In a hypothetical non-multi-tenant caller, the engine could still be reusable global thing. For the mailroom multi-tenant case, we'd probably instantiate an engine same way we do for an environment, e.g.

1. Look at org to see if it has a email configuration (it has an SMTP configuration)
2. Instantiate an SMTP email service
3. Instantiate an engine with that
4. Start sessions with that engine instance

Or maybe it's cached the way assets are? Can't see it being expensive to instantiate an engine so maybe it doesn't matter?

**Comment by nicpottier** [comment]
I'm not sure these are either org or env specific in reality, though they may refer to variables on the org or asset.

For example, for NLP, if we say we have assets that are Models, then the interface is probably something like:
  `Classify(env, model, language, text)`

And model may be a reference either to a wit.ai model or a bothub model, but those can be different backends depending on the model.

So that leads me more to the path that providers are passed into the engine creation and the calls to providers pass in all the state that is needed to perform the action in a stateless way.

Note that the other consideration here is logging, seems all of these probably need some way of logging what the provider did, probably as a set of request/responses?

**Comment by rowanseymour** [comment]
Ok so yeah another way to approach this is that there's not a `SMTPEmailService` and a `SESEmailService` but a single `EmailService` which goflow delegates email sending to, and it is passed some state which allows it to find the configuration for sending an email and use the appropriate service depending on what has been configured.

It all looks the same to flow elements in goflow, but engine can still be global reusable thing in mailroom.

I was thinking about this approach previously and how if you just passed in the flow to the service function then mailroom could derive the org from the flow UUID and lookup the org's email configuration. But maybe it would be cleaner to add something to environment which mailroom can use to pass around org ids? Or pass it the whole session?

> And model may be a reference either to a wit.ai model or a bothub model, but those can be different backends depending on the model. (and importantly a single org may have models across both)

I think it's worth pausing for a second and asking ourselves if we really need to support multiple services for the same thing within the same org, but if the answer is yes then I think NLP/NLU models have to be assets and they're another dependency type for a flow.

A classify action would have a reference to a NLP/NLU model asset and would pass it back to the "NLP service" something like

```golang
model := run.Session().Assets().NLPModels().Get(a.Model.UUID)
if model == nil {
   // error event?
}  

value := run.EvaluateTemplate(a.Value)

category, event := run.Session().Engine().NLP().Classify(env, model, value)

logEvent(event)

// save result
run.SaveResult(flows.NewResult(value, category, ...))
```

Re logging.. for every "service" there could be a corresponding event type, so in this case, mailroom's NLP classify method has to return a `nlp_classified` event which gets logged in the session.

**Comment by nicpottier** [comment]
I don't see any way Models *can't* be assets. There is more than one model for sure (even if on the same provider) and the editor will need to say which model to use, so doesn't that mean Models are then assets? I don't think we want to have people enter some sort of arbitrary id in there do we?

This does create a dependency but I don't see what choice we have. The action can still execute but go down an error path if the model doesn't exist? (which needs to be handled either way)

Agree on the event stuff.

**Comment by rowanseymour** [comment]
Yeah I was some what hoping in vain for something that wasn't another dependency we need to track on the RP side. Other argument for these being treated as hard dependencies is that we really need the model to continue the flow. It's not like a add_groups action where we can log a warning and move on if the group doesn't exist. The router below the classify action will need a result to route on.

**Comment by nicpottier** [comment]
Ya, though it seems since this is a split there will be an error case and the model missing would be one of those errors.

**Comment by rowanseymour** [comment]
True what with this needing a call to an external thing (like `call_webhook`) that may or may not succeed.

Ok last thing to think about. If an action calls an implementation of `Classify(env, model, value)` then that implementation (i.e. mailroom) can easily infer from the `model` asset which org it's dealing with, get the config etc.

In the case of `Emails().Send(env, subject, body)` there isn't anything for the mailroom to know what org it's dealing with and how it sends email.

If these methods passed back the flow then mailroom could lookup org based on the flow UUID but that's a db hit.
